### PR TITLE
treestatus: implement Treestatus UI (Bug 1817473)

### PR DESCRIPTION
**Pull request by cgsheeh**
Implement a Treestatus UI in Lando. All endpoints in the new
interface are namespaced under `/treestatus`. The main components
are as follows.

The main Treestatus page displays all trees and their current statuses.
When authenticated, each tree can be selected for updating via a checkbox,
or a "select all trees" button is available for CI-wide tree closures.
The tree selections are forwarded to the update form when "Update trees"
is selected.

The "update trees" view implements a form for updating the status of
trees. Trees can be set to open, closed and "approval required". When
the status is being set to a non-open state, a reason and reason
category must be specified. The form presents a "remember this change"
option which can be used to easily revert the change at a later time.
For example if all trees are closed, the trees can be mass re-opened
easily.

The individual log view of a tree presents the most recent status
updates for each individual tree. When authenticated, users can edit
the tree status reason or category in case of a typo or error during
submission by clicking the "edit" button, which toggles the display to a
form for updating.

The recent changes view is a header that appears for authenticated users
on all pages. It presents all status updates that have been "remembered"
during update. Users can edit the reason and reason category using an
"edit" button that toggles the display to a form. The "restore" button
can be used to revert the status change. The "discard" button can be
used to throw away the status change if it is no longer necessary to
remember the change.

The Treestatus UI uses message flashing and the existing Lando-UI
notifications component to present errors and action verification to the
user.

### CODE REVIEW

**Comment by cgsheeh** [comment]
I still need to demo this UI to the sheriffs and make sure I got all the requirements, and also chat with the ops team about how to migrate the DB over (set up staging or dev for testing as well). I wanted to get this up sooner than later in case there's anything glaring that I missed or that needs to be changed. :)

Some TODOs that I'm already aware of:

- there should be a link to Treestatus from the main Lando page.
- Implement a "delete tree" button on the main page and a confirmation form.
- It would be nice to show the current status of the trees when updating to a new status (ie "this tree is open and you're about to make it closed").

**Comment by cgsheeh** [comment]
Screenshots:

log_authed
main_page_authed
main_page_unauthed
update_trees

**Comment by cgsheeh** [comment]
I demoed the new UI to Aryx, we identified a few more things that should be handled before this lands:

- "remember this change" should default to selected in the UI.
- Trees should be sorted by relevance in the tree selection display.
- The cards in the tree select view should be a little smaller/spaced closer together to reduce the amount of scrolling needed to select trees.
- The values of tags that are passed to the backend need to match up with the current state. Right now we pass the display strings like "Check-in compilation failure", but the value that should end up in the backend should be something like `checkin_compilation`. We will need to map between these two values.

**Review by zzzeid — CHANGES_REQUESTED** [review]
First pass.

In the title: `feat: ` -> `treestatus: `

My comments apply to various parts of the code, but I did not leave them in every applicable part so please keep that in mind. The main overarching comment is that the "Web UI" functionality should be better separated from the "backend" functionality, and moving to an async approach would make the code cleaner (in terms of separation/gouping of functionality, status codes, testing, etc...) and easier to maintain in the future. Some of my comments however can be addressed without making this change.

There's also no test coverage for treestatus.py.

File: landoui/treestatus.py
Comment (cgsheeh) [reply landoui/treestatus.py:85]: Wrote a PR for this since it's a problem that's pre-existing in the code. See #189 .

File: landoui/treestatus.py
Comment (cgsheeh) [reply landoui/treestatus.py:163]: `getattr` is only really appropriate to use when the name of the attribute isn't known ahead of time, or when some other logic is used to determine which attribute should be accessed. When we know the name of the variable it's better to explicitly use it. Using `getattr` in this way breaks LSP renaming, "get all references", etc.

For example, I will use it to rename `remember_this_change` to just `remember` to match what the API is expecting in the JSON body, heh. :sweat_smile: 

We can still clean up the code in other ways, perhaps with a `TreeStatusUpdateTreesForm.to_json()` method or something of that nature.

File: tests/test_treestatus.py
Comment (zzzeid) [inline tests/test_treestatus.py:92]: I guess the assertion here is that no exception will be raised? Wondering if the explanation of failure will show if an exception is raised (given that it's not an inequality that will be assessed.) Just not sure of the behaviour, but curious to know.

File: landoui/treestatus.py
Comment (zzzeid) [inline landoui/treestatus.py:79]: What is "Update trees"? Maybe be more explicit here (e.g., "submitting the update tree form").

**Review by zzzeid — APPROVED** [review]
Looks good to me! Just a couple of small comments.

File: landoui/treestatus.py
Comment (cgsheeh) [reply landoui/treestatus.py:79]: "Update trees" is the text on the button used to display the form modal. I'll change this to `clicking "Update trees"`.

File: tests/test_treestatus.py
Comment (zzzeid) [inline tests/test_treestatus.py:103]: nit: Might be cleaner to just do
```suggestion
        raise AssertionError(f"No validation required for `open` status: {exc}.")
```

**Review by zzzeid — APPROVED** [review]

**Review by zzzeid — CHANGES_REQUESTED** [review]

File: landoui/landoapi.py
Comment (cgsheeh) [reply landoui/landoapi.py:134]: A parameter passed to the class would also have been an acceptable approach, but I don't see anything wrong with a separate class in this instance. I think I'm going to leave this as-is, it seems like more work than it's worth to change now, especially with the added context that we will be removing all of this code in Django-Lando anyways.

Comment (zzzeid) [reply landoui/landoapi.py:134]: This approach adds more complexity and technical debt than is needed. I.e., separate "service names", as well as separate classes, which takes away from the fact that the two classes are both "Lando API" wrappers wrapping the same API. It may be easier to visualize my point by picturing what this will look like in the unified Lando project -- there is no concept of `LandoAPI` and `TreestatusAPI` -- everything is in the same project and is directly called from within the project itself.

**Review by zzzeid — APPROVED** [review]
I won't block this PR on my last comment (added my response directly there), as this is something you will be able to clean up when these changes are moved to the Lando repo.

**Pull request merged**
